### PR TITLE
wip implement AD for tensors

### PR DESCRIPTION
**Pull request by KristofferC**
Just some very small testing for vectors.

```jl
julia> v = rand(Vec{3})
3-element ContMechTensors.Tensor{1,3,Float64,3}:
 0.268185
 0.662525
 0.265921

julia> ContMechTensors.gradient(norm, v)
3-element ContMechTensors.Tensor{1,3,Float64,3}:
 0.351666
 0.868758
 0.348698

julia> v / norm(v)
3-element ContMechTensors.Tensor{1,3,Float64,3}:
 0.351666
 0.868758
 0.348698

julia> @code_llvm ContMechTensors.gradient(norm, v)

define void @julia_gradient_73321(%Tensor* noalias sret, %Tensor*) #0 {
pass:
  %2 = getelementptr inbounds %Tensor, %Tensor* %1, i64 0, i32 0, i32 0, i64 0
  %3 = load double, double* %2, align 1
  %4 = getelementptr inbounds %Tensor, %Tensor* %1, i64 0, i32 0, i32 0, i64 1
  %5 = load double, double* %4, align 1
  %6 = getelementptr inbounds %Tensor, %Tensor* %1, i64 0, i32 0, i32 0, i64 2
  %7 = load double, double* %6, align 1
  %8 = fmul double %3, %3
  %9 = fmul double %5, %5
  %10 = fadd double %8, %9
  %11 = fmul double %7, %7
  %12 = fadd double %10, %11
  %13 = fmul double %3, 0.000000e+00
  %14 = fadd double %13, %13
  %15 = fmul double %5, 0.000000e+00
  %16 = fadd double %15, %15
  %17 = fadd double %14, %16
  %18 = fadd double %7, %7
  %19 = fadd double %18, %17
  %20 = fadd double %5, %5
  %21 = fadd double %14, %20
  %22 = fmul double %7, 0.000000e+00
  %23 = fadd double %22, %22
  %24 = fadd double %21, %23
  %25 = fadd double %3, %3
  %26 = fadd double %25, %16
  %27 = fadd double %26, %23
  %28 = call double @llvm.sqrt.f64(double %12)
  %29 = fdiv double 5.000000e-01, %28
  %30 = fmul double %27, %29
  %31 = fmul double %24, %29
  %32 = fmul double %19, %29
  %33 = getelementptr inbounds %Tensor, %Tensor* %0, i64 0, i32 0, i32 0, i64 0
  store double %30, double* %33, align 8
  %34 = getelementptr inbounds %Tensor, %Tensor* %0, i64 0, i32 0, i32 0, i64 1
  store double %31, double* %34, align 8
  %35 = getelementptr inbounds %Tensor, %Tensor* %0, i64 0, i32 0, i32 0, i64 2
  store double %32, double* %35, align 8
  ret void
}
```

### CODE REVIEW

**Comment by KristofferC** [comment]
accidentally added some voigt crap... I remove it later.

**Comment by KristofferC** [comment]
For Symmetric tensors, isn't it just enough to load the offdiagonals with an inital 0.5 instead of 1 for the dual seeds!??!

**Comment by fredrikekre** [comment]
The test commit shouldn't be included here I assume?

Otherwise it looks nice, this is a really good functionality to support :)

**Comment by KristofferC** [comment]
It is here because it was annoying to comment out the tests to run only the AD tests. When my other PR is merged I rebase this one and the test commit will disappear.

**Comment by KristofferC** [comment]
Starting to get some testing done now :)

Can you figure this out https://github.com/KristofferC/ContMechTensors.jl/pull/86/commits/07253b62b2b67ec587386fd1108aecc8342d9c0d#diff-98bb40dafe56e2e7215dd38e4895b5e1R56

**Comment by fredrikekre** [comment]
> Can you figure this out

IIRC `S` is defined as `S = 2 * ∂Ψ∂C`

**Comment by KristofferC** [comment]
Just figured it out and rushed to my computer to delete it before you had time to see it :(

**Comment by fredrikekre** [comment]
```jl
julia> @using ContMechTensors as cmt

julia> graddet(A) = det(A) * inv(A)';

julia> A = rand(Tensor{2,3});

julia> @benchmark cmt.gradient($det, $A)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     18.579 ns (0.00% GC)
  median time:      18.615 ns (0.00% GC)
  mean time:        19.846 ns (0.00% GC)
  maximum time:     68.949 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     996
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark graddet($A)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     23.301 ns (0.00% GC)
  median time:      23.354 ns (0.00% GC)
  mean time:        24.460 ns (0.00% GC)
  maximum time:     73.187 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     990
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by KristofferC** [comment]
Implementing hessians. Much difficult... Much code.

File: test/test_ad.jl [hunk 1]
```diff
@@ -1,4 +1,7 @@
- function Ψ(C, μ, Kb)
+const ∇ = ContMechTensors.gradient
+const Δ = ContMechTensors.hessian
```
Comment (fredrikekre) [inline test/test_ad.jl:2]:
```jl
const ∇² = ContMechTensors.hessian
```

**Comment by KristofferC** [comment]
Rebased on master.

**Comment by KristofferC** [comment]
The Hessians will always be symmetric?

**Comment by fredrikekre** [comment]
Yes.
```jl
df/dxdy == df/dydx
```

**Comment by KristofferC** [comment]
Also, shouldnt this be equal?
```jl
julia> srand(1234)

julia> ψ(ϵ, E) = 1/2 * ϵ ⊡ E ⊡ ϵ;

julia> ϵ = rand(SymmetricTensor{2, 2})

julia> E = rand(SymmetricTensor{4, 2})

julia> E2 = ContMechTensors.hessian(ϵ ->  ψ(ϵ, E), ϵ)

julia> E - E2
2×2×2×2 ContMechTensors.SymmetricTensor{4,2,Float64,9}:
[:, :, 1, 1] =
 0.0      0.29672 
 0.29672  0.137237

[:, :, 2, 1] =
 -0.29672   0.0     
  0.0      -0.201022

[:, :, 1, 2] =
 -0.29672   0.0     
  0.0      -0.201022

[:, :, 2, 2] =
 -0.137237  0.201022
  0.201022  0.0  
```

**Comment by KristofferC** [comment]
Symmetric in terms of major transpose then apparently:

```jl
julia> H = ContMechTensors.hessian(norm, rand(Tensor{2,2}));

julia> H - H'
2×2×2×2 ContMechTensors.Tensor{4,2,Float64,16}:
[:, :, 1, 1] =
  0.0        0.0131206
 -0.0131206  0.0      

[:, :, 2, 1] =
 -0.0131206   0.0      
 -0.316729   -0.0222194

[:, :, 1, 2] =
 0.0131206  0.316729 
 0.0        0.0222194

[:, :, 2, 2] =
  0.0        0.0222194
 -0.0222194  0.0      

julia> H - majortranspose(H)
2×2×2×2 ContMechTensors.Tensor{4,2,Float64,16}:
[:, :, 1, 1] =
  0.0          -3.46945e-18
 -6.93889e-18  -8.67362e-19

[:, :, 2, 1] =
 6.93889e-18  0.0
 0.0          0.0

[:, :, 1, 2] =
 3.46945e-18  0.0
 0.0          0.0

[:, :, 2, 2] =
 8.67362e-19  0.0
 0.0          0.0
```

**Comment by fredrikekre** [comment]
Do we need to seed with `one(T) / 4` then for some indices?

**Comment by fredrikekre** [comment]
That might give minor symmetry as well.

**Comment by KristofferC** [comment]
I got different answer when using full tensors as well... hmm.

**Comment by KristofferC** [comment]
Because I am stupid.. for full tensors you need to take the symmetric part:

```jl
ψ2(ϵ, E) = 1/2 * dot(ϵ, E * ϵ);
ϵ2 = rand(4);
E = rand(4, 4);
ForwardDiff.hessian(ϵ ->  ψ2(ϵ, E), ϵ2)

ψ(ϵ, E) = 1/2 * ϵ ⊡ E ⊡ ϵ;
ϵ = Tensor{2,2}(ϵ2);
E = Tensor{4,2}(E);
ContMechTensors.hessian(ϵ ->  ψ(ϵ, E), ϵ)
```

**Comment by KristofferC** [comment]
But yeah, maybe have to do the 1/4 like u say for symmetric tensors.

**Comment by KristofferC** [comment]
Added some docs.

File: docs/make.jl [hunk 2]
```diff
@@ -4,6 +4,7 @@ makedocs(
     modules = [ContMechTensors],
     format = :html,
     sitename = "ContMechTensors.jl",
+    doctest = true,
```
Comment (fredrikekre) [inline docs/make.jl:7]: Did this default to true? doctests were enabled before too

File: docs/src/demos.md [hunk 3]
```diff
@@ -29,13 +29,13 @@ C = SymmetricTensor{4, dim}(f)
...
 
-We can define the energy for a material with the
+We can define potential energy of the material as
 
 $\Psi(\mathbf{C}) = 1/2 \mu (\mathrm{tr}(\hat{\mathbf{C}}) - 3) + K_b(J-1)^2,$
 
 where $\hat{\mathbf{C}} = \mathrm{det}(\mathbf{C})^{-1/3} \mathbf{C}$ and $J = \det(\mathbf{F}) = \sqrt{\det(\mathbf{C})}$ and the shear and bulk modulus are given by $\mu$ and $K_b$ respectively.
 
-This free energy function can be implemented as:
+This free energy function can be implemented in `ContMechTensors as:
```
Comment (fredrikekre) [inline docs/src/demos.md]: missing backtick

File: docs/src/man/automatic_differentiation.md [hunk 4]
```diff
@@ -1,7 +1,71 @@
...
+DocTestSetup = quote
+    srand(1234)
+    using ContMechTensors
+end
+```
+
 # Automatic Differentiation
 
-`ContMechTensors` supports automatic differentiation of tensorial functions to compute the gradients and Hessian's of functions.
+`ContMechTensors` supports automatic differentiation (AD) of tensorial functions to compute first order derivatives (gradients) and second order derivatives (Hessians).
```
Comment (fredrikekre) [inline docs/src/man/automatic_differentiation.md]: mention that it is forward mode AD?

File: docs/src/man/automatic_differentiation.md [hunk 5]
```diff
@@ -1,7 +1,71 @@
...
+It does this by exploiting the `Dual` number defined in `ForwardDiff.jl`.
+While `ForwardDiff.jl` can itself be used to differentiate tensor functions it is a bit awkward because `ForwardDiff.jl` is written to work with standard Julia `Array`s. One therefore has to send the input argument as an `Array` to `ForwardDiff.jl`, convert it to a `Tensor` and then convert the output `Array` to a `Tensor` again. This can also be inefficient since these `Array`s are allocated on the heap so one needs to preallocte which can be annoying.
+
+Instead, it is simpler to use `ContMechTensors` own AD API to do the differentiation. This does not require any conversions and everything will be stack allocated so there is no need to preallocate.
+
+The API for AD in `ContMechTensors` is `ContMechTensors.gradient(f, A)` and ``ContMechTensors.hessian(f, A)` where `f` is a function and `A` is a first or second order tensor. For `gradient` the function can return a scalar, vector (in case the input is a vector) or a second order tensor. For `hessian` the function should return a scalar.
```
Comment (fredrikekre) [inline docs/src/man/automatic_differentiation.md]: double backtic

File: docs/src/man/automatic_differentiation.md [hunk 6]
```diff
@@ -1,7 +1,71 @@
...
+
+```jldoctest
+julia> x = rand(Vec{2});
+
+julia> ContMechTensors.gradient(norm, x)
+2-element ContMechTensors.Tensor{1,2,Float64,2}:
+ 0.61036
+ 0.792124
+
+julia> x / norm(x)
```
Comment (fredrikekre) [inline docs/src/man/automatic_differentiation.md:32]: should mention that this is the analytical solution

File: docs/src/man/automatic_differentiation.md [hunk 7]
```diff
@@ -1,7 +1,71 @@
...
+
+```jldoctest
+julia> A = rand(SymmetricTensor{2,2});
+
+julia> ContMechTensors.gradient(det, A)
+2×2 ContMechTensors.SymmetricTensor{2,2,Float64,3}:
+  0.566237  -0.766797
+ -0.766797   0.590845
+
+julia> inv(A)' * det(A)
```
Comment (fredrikekre) [inline docs/src/man/automatic_differentiation.md:50]: mention analytical solution here too

File: docs/src/man/automatic_differentiation.md (hunk 6 above)
Comment (KristofferC) [reply docs/src/man/automatic_differentiation.md:32]: I do in the math above?

File: docs/src/man/automatic_differentiation.md (hunk 7 above)
Comment (KristofferC) [reply docs/src/man/automatic_differentiation.md:50]: Same here

File: docs/src/man/automatic_differentiation.md (hunk 6 above)
Comment (fredrikekre) [reply docs/src/man/automatic_differentiation.md:32]: Oh

File: docs/make.jl (hunk 2 above)
Comment (KristofferC) [reply docs/make.jl:7]: Yes, but I wanted to turn them off for a while so I had it to false. It doesn't hurt to be explicit.

Comment (fredrikekre) [reply docs/make.jl:7]: Just got confused

**Comment by KristofferC** [comment]
Some more tests. It is a bit tricky when you take the symmetric derivative because of the non uniqueness of the skew component. I therefore tested that the answer is the same when multiplying with a symmetric second order tensor.

**Comment by KristofferC** [comment]
If tests pass, this should be good to go from my point of view.

**Comment by fredrikekre** [comment]
Am I missing something?

```jl
julia> @using ContMechTensors as cmt

julia> Ψ = e -> 1/2 * e ⊡ E ⊡ e;

julia> ϵ = rand(SymmetricTensor{2,3});

julia> E = rand(SymmetricTensor{4,3});

julia> σ_ad = cmt.gradient(Ψ, ϵ)
3×3 ContMechTensors.SymmetricTensor{2,3,Float64,6}:
 3.37388  3.97431  2.56796
 3.97431  2.82679  3.60384
 2.56796  3.60384  3.30445

julia> σ = E ⊡ ϵ
3×3 ContMechTensors.SymmetricTensor{2,3,Float64,6}:
 4.39254  3.73682  2.18276
 3.73682  3.05402  3.89903
 2.18276  3.89903  3.56543
```

**Comment by fredrikekre** [comment]
```jl
julia> Ψ2 = e -> 1/2 * e ⊡ majorsymmetric(E) ⊡ e

julia> σ_ad2 = cmt.gradient(Ψ2, ϵ)
3×3 ContMechTensors.SymmetricTensor{2,3,Float64,6}:
 3.37388  3.97431  2.56796
 3.97431  2.82679  3.60384
 2.56796  3.60384  3.30445

julia> majorsymmetric(E) ⊡ ϵ
3×3 ContMechTensors.Tensor{2,3,Float64,9}:
 3.37388  3.97431  2.56796
 3.97431  2.82679  3.60384
 2.56796  3.60384  3.30445
```

**Comment by fredrikekre** [comment]
Properly done:

```jl
julia> @using ContMechTensors as cmt

julia> μ = 1.0; λ = 2.0;

julia> Is = one(SymmetricTensor{2,3}); IIs = one(SymmetricTensor{4,3});

julia> C = λ * Is ⊗ Is + 2*μ*IIs;

julia> isminorsymmetric(C)
true

julia> ismajorsymmetric(C)
true

julia> Ψ = e -> 1/2 * e ⊡ C ⊡ e;

julia> ϵ = rand(SymmetricTensor{2,3});

julia> σ_ad = cmt.gradient(Ψ, ϵ)
3×3 ContMechTensors.SymmetricTensor{2,3,Float64,6}:
 1.73452   0.515906  1.27075 
 0.515906  1.45422   0.852944
 1.27075   0.852944  1.08977 

julia> σ = C ⊡ ϵ
3×3 ContMechTensors.SymmetricTensor{2,3,Float64,6}:
 1.73452   0.515906  1.27075 
 0.515906  1.45422   0.852944
 1.27075   0.852944  1.08977 

julia> norm(σ-σ_ad)
0.0

julia> C_ad = cmt.hessian(Ψ, ϵ);

julia> norm(C - C_ad)
0.0
```

**Comment by fredrikekre** [comment]
(For future updates of this we should have benchmarks too, doesn't matter now since we don't have much to compare with though.)

File: README.md [hunk 8]
```diff
@@ -9,9 +9,10 @@
 ## Introduction
 
 This Julia package provides fast operations with symmetric and non-symmetric tensors of order 1, 2 and 4.
-The Tensors are allocated on the stack which means that there is no need to preallocate output results for performance. 
+The Tensors are allocated on the stack which means that there is no need to preallocate output results for performance.
```
Comment (fredrikekre) [inline README.md:12]: sublime has an option to remove trailing whitespace on save, pretty nice

**Comment by KristofferC** [comment]
I agree. I make an issue about it. We can always compare to the analytical ones. I can work on it now but doesnt have to be same PR.

File: README.md (hunk 8 above)
Comment (KristofferC) [reply README.md:12]: Yeah, have it installed on work computer but need to fix it everywhere.

**Comment by KristofferC** [comment]
Why isnt coverage popping up?

**Comment by fredrikekre** [comment]
Codecov only comments once per PR.

But the report should be able to be seen [here](https://codecov.io/gh/KristofferC/ContMechTensors.jl/pull/86)

But is says
screenshot from 2016-12-10 15-26-57

no `base report` to compare with for some reason? Idk...

**Comment by KristofferC** [comment]
Let's do this? I add tests if coverage is whining later.

**Comment by fredrikekre** [comment]
Go ahead!

**Pull request merged**